### PR TITLE
[#99] Add support for changing passwords (4-2-stable)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: irods_specific_implementation/irods_rest_api_base.h [hunk 1]
```diff
@@ -232,7 +232,7 @@ namespace irods::rest
 
         std::string decode_url(const std::string& _in) const
         {
-            trace("Decoding input [{}] ...", _in);
+            trace("Decoding input [{}] ...", _in); // TODO This prints incoming arguments! (e.g. Passwords!)
```
Comment (trel) [inline irods_specific_implementation/irods_rest_api_base.h:235]: oh.   hmmm...  yeah i don't think we should have this here.   debugging is easy enough - consider commenting it out, so developers can easily turn it back on?

**Review by trel — COMMENTED** [review]
do we need a test?

File: README.md [hunk 2]
```diff
@@ -145,7 +145,7 @@ An iRODS ticket token within the **irods-ticket** header, and a URL for streamin
 ```
 
 ### /admin
-The administration interface to the iRODS Catalog which allows the creation, removal and modification of users, groups, resources, and other entities within the zone.
+The administration interface to the iRODS Catalog which allows the creation, removal and modification of users, groups, resources, and other entities within the zone. **ALL input parameters must be defined! If an input parameter is not to be used, set it to nothing (e.g. arg7=)**
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply README.md:148]: we can, but the `rxGeneralAdmin` API's input structure uses generic names for member variables and we don't know what arguments will be used by the server unless we check them.
```c++
struct input
{
    void* arg0;
    void* arg1;
    ...
    void* arg7;
};
```
not only that, but the server may be sensitive to empty strings and null values.

we can default to empty strings and pass null to the rc call, but how do we determine if the client meant to send an empty string?

File: irods_specific_implementation/irods_rest_admin_api_implementation.h [hunk 3]
```diff
@@ -73,6 +93,38 @@ namespace irods::rest
...
+
+            std::array<char, MAX_PASSWORD_LEN + 10> admin_password{};
+
+            // Get the plain text password of the iRODS service account user.
+            // "obfGetPw" decodes the obfuscated password stored in .irods/.irodsA.
+            if (obfGetPw(admin_password.data()) != 0) {
+                THROW(SYS_INTERNAL_ERR, "failed to unobfuscate admin password in .irodsA file.");
+            }
+
+            debug("Client session signature = [{}]", getSessionSignatureClientside());
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h:116]: i'll remove it. it doesn't seem very useful to me.

File: irods_specific_implementation/irods_rest_api_base.h (hunk 1 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_api_base.h]: left comment explaining why it is commented out.

File: irods_specific_implementation/irods_rest_admin_api_implementation.h [hunk 4]
```diff
@@ -73,6 +93,38 @@ namespace irods::rest
...
+
+            std::array<char, MAX_PASSWORD_LEN + 100> obfuscate_password{};
+            obfEncodeByKeyV2(plain_text_password.data(),
+                             admin_password.data(),
+                             getSessionSignatureClientside(),
+                             obfuscate_password.data());
+
+            debug("Obfuscated password = [{}]", std::string(obfuscate_password.data()));
+
+            return obfuscate_password.data();
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h]: done

File: irods_specific_implementation/irods_rest_admin_api_implementation.h [hunk 5]
```diff
@@ -73,6 +93,38 @@ namespace irods::rest
...
+
+            debug("Client session signature = [{}]", getSessionSignatureClientside());
+
+            std::array<char, MAX_PASSWORD_LEN + 100> obfuscate_password{};
+            obfEncodeByKeyV2(plain_text_password.data(),
+                             admin_password.data(),
+                             getSessionSignatureClientside(),
+                             obfuscate_password.data());
+
+            debug("Obfuscated password = [{}]", std::string(obfuscate_password.data()));
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h]: done

File: irods_specific_implementation/irods_rest_admin_api_implementation.h [hunk 6]
```diff
@@ -73,6 +93,38 @@ namespace irods::rest
...
+
+            // Get the plain text password of the iRODS service account user.
+            // "obfGetPw" decodes the obfuscated password stored in .irods/.irodsA.
+            if (obfGetPw(admin_password.data()) != 0) {
+                THROW(SYS_INTERNAL_ERR, "failed to unobfuscate admin password in .irodsA file.");
+            }
+
+            debug("Client session signature = [{}]", getSessionSignatureClientside());
+
+            std::array<char, MAX_PASSWORD_LEN + 100> obfuscate_password{};
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h]: done

File: irods_specific_implementation/irods_rest_admin_api_implementation.h [hunk 7]
```diff
@@ -73,6 +93,38 @@ namespace irods::rest
...
+            obfEncodeByKeyV2(plain_text_password.data(),
+                             admin_password.data(),
+                             getSessionSignatureClientside(),
+                             obfuscate_password.data());
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h]: done

File: irods_specific_implementation/irods_rest_admin_api_implementation.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h]: done

**Comment by korydraughn** [comment]
im planning on adding a test very soon.

File: irods_specific_implementation/irods_rest_admin_api_implementation.h [hunk 8]
```diff
@@ -73,6 +93,36 @@ namespace irods::rest
...
+                THROW(SYS_INTERNAL_ERR, "failed to unobfuscate admin password in .irodsA file.");
+            }
+
+            std::array<char, MAX_PASSWORD_LEN + 100> obfuscated_password{};
+            obfEncodeByKeyV2(plain_text_password.data(),
+                             admin_password.data(),
+                             getSessionSignatureClientside(),
+                             obfuscated_password.data());
+
+            debug("Obfuscated password = [{}]", std::string(obfuscated_password.data()));
```
Comment (trel) [inline irods_specific_implementation/irods_rest_admin_api_implementation.h:122]: remove line

Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h]: done

**Comment by korydraughn** [comment]
added test.

all tests pass at the bench.

**Comment by korydraughn** [comment]
added section to README about plaintext passwords and HTTPS.

please review.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (trel) [reply README.md:153]: agreed.  requiring all the parameters seems best for now.

File: packaging/irods_rest.py [hunk 9]
```diff
@@ -217,7 +217,7 @@ def admin(_token, _action, _target, _arg2, _arg3, _arg4, _arg5, _arg6, _arg7):
     c = pycurl.Curl()
     c.setopt(pycurl.HTTPHEADER,['Accept: application/json'])
     c.setopt(pycurl.HTTPHEADER,['Authorization: '+_token])
-    c.setopt(c.CUSTOMREQUEST, 'GET')
+    c.setopt(c.CUSTOMREQUEST, 'POST')
```
Comment (trel) [inline packaging/irods_rest.py:220]: this seems really important - but also... seems more correct.    why/how was it GET before?

Comment (korydraughn) [reply packaging/irods_rest.py:220]: was never used in a test before now.

File: packaging/test_irods_client_rest_cpp.py [hunk 10]
```diff
@@ -452,3 +453,19 @@ def test_zone_report(self):
 
             assert(js0 == js1)
 
+    def test_changing_passwords_is_supported__issue_99(self):
+        # Show that the user can execute commands without error.
+        self.user.assert_icommand(['ils', '-ld'], 'STDOUT', [self.user.session_collection])
+
+        # Show that changing the user's password will require them to reauthenticate.
+        token = irods_rest.authenticate('rods', 'rods', 'native')
+        new_password = 'rods'
```
Comment (trel) [inline packaging/test_irods_client_rest_cpp.py:463]: new password same as the old?

File: packaging/test_irods_client_rest_cpp.py [hunk 11]
```diff
@@ -452,3 +453,19 @@ def test_zone_report(self):
...
+        irods_rest.admin(token, 'modify', 'user', self.user.username, 'password', new_password, None, None, None)
```
Comment (trel) [inline packaging/test_irods_client_rest_cpp.py:464]: so this is the admin changing alice's password?

File: packaging/irods_rest.py (hunk 9 above)
Comment (trel) [reply packaging/irods_rest.py:220]: gotcha

File: packaging/test_irods_client_rest_cpp.py (hunk 10 above)
Comment (korydraughn) [reply packaging/test_irods_client_rest_cpp.py:463]: that is the new password for the user inherited by the test.

the user's original password is 'apass'. see class name.

File: packaging/test_irods_client_rest_cpp.py (hunk 11 above)
Comment (korydraughn) [reply packaging/test_irods_client_rest_cpp.py:464]: yes

**Comment by korydraughn** [comment]
are we happy with the exclamation marks in commit messages, etc. and the HTTPS section in the README?

File: packaging/test_irods_client_rest_cpp.py (hunk 10 above)
Comment (trel) [reply packaging/test_irods_client_rest_cpp.py:463]: oh, so let's change it to be something that is unique from the default admin password.     make it 'apassnew' or something that is self-describing.

**Comment by trel** [comment]
i tend to lean away from any emotion in technical documentation.  generally, consider that readers could be english-as-second/third-language - so want to be very clear with the words/images (no metaphors or cultural references), and not lean on punctuation as much.

**Comment by trel** [comment]
yeps - looks good - please # it.

**Comment by korydraughn** [comment]
added #

**Pull request merged**